src/Development/IDE/Plugin/CodeAction/Rules.hs

```diff
@@ -0,0 +1,59 @@
+module Development.IDE.Plugin.CodeAction.Rules
+  ( rulePackageExports,
+    getVisiblePackageExports
+  )
+where
+
+import Data.HashMap.Strict (fromListWith)
+import qualified Data.HashMap.Strict as Map
+import Data.Text ( Text, pack,)
+import Data.Traversable (forM)
+import Development.IDE.Core.RuleTypes
+import Development.IDE.Core.Shake
+import Development.IDE.GHC.Util
+import Development.IDE.Plugin.CodeAction.RuleTypes
+import Development.IDE.Types.Location
+import Development.Shake
+import HscTypes (hsc_dflags, IfaceExport, mi_exports)
+import LoadIface
+import Maybes
+import Module (Module (..), ModuleName, moduleNameString)
+import Packages (explicitPackages, exposedModules, packageConfigId)
+import TcRnMonad (WhereFrom (ImportByUser), initIfaceLoad)
+import GHC (DynFlags(pkgState))
+import Control.Monad.Trans (MonadTrans(lift))
+
+getVisiblePackageExports :: NormalizedFilePath -> Action [PackageExportsMap]
+getVisiblePackageExports f = fmap (fromMaybe []) $ runMaybeT $ do
```

Comment:
This looks unnecessarily expensive. Afaict you are calling this for individual files but it only depends on the GHC session which is the same at least within a component. Can we only do this once per GHC session?

Reply:
Why do you think this is expensive? It's just retrieving things from the Shake graph, there is no actual computation involved.

If you are concerned it can be cached in the Shake graph, at the cost of more memory.

Reply:
Sorry should have been more clear, I mean the `rulePackageExports` rule which seems to be called once per `dependencies * files` in your project.

Reply:
You are right, the `PackageExports` rule is file-dependent. The suggestion is to make it session dependent instead. But that seems tricky, since a session is not something that can be used as a key in the Shake graph. Will have to think about it

Reply:
`HscEnvEq` can probably be used as a key so you might be able to define it with `defineNoFile` and accept the `HscEnvEq` as an argument. Alternatively something like the `memoIO` approach used to initialize the session itself might work but that might be a bit tricky given that the internals of `ghcide` don’t know about this.

Reply:
Thanks for the suggestion to go with `HscEnvEq`.